gloo/transport/tcp/context.cc

```diff
@@ -192,7 +192,10 @@ bool Context::findRecvFromAny(
 }
 
 void Context::signalException(const std::string& msg) {
-  std::unique_lock<std::mutex> lock(m_);
+  // The `pairs_` vector is logically constant. After the context and
+  // all of its pairs have been created it is not mutated until the
+  // context is destructed. Therefore, we don't need to acquire this
+  // context's instance lock before looping over `pairs_`.
   for (auto& pair : pairs_) {
```

Comment:
Will there be a problem if `createPair` has not been called on all ranks? Will those late ranks miss the exception signal?

Reply:
The creation of these pairs is done before any operation is run against them, so this can't happen. In order to protect against it, keeping the mutex around wouldn't be enough either, but would require adding some kind of state variable that tells us whether or not initialization is complete.